File: csharp/build-company/StructuralMaterial.cs
namespace BuildCompany;

/// <summary>
/// Класс для конструкционных материалов (кирпичи, балки, панели и т.д.)
/// </summary>
public class StructuralMaterial : Material
{
    private float width;
    private float height;
    private float length;
    private int count;

    /// <summary>
    /// Конструктор
    /// </summary>
    /// <param name="name">Название материала</param>
    /// <param name="width">Ширина в метрах</param>
    /// <param name="height">Высота в метрах</param>
    /// <param name="length">Длина в метрах</param>
    /// <param name="count">Количество</param>
    public StructuralMaterial(string name, float width, float height, float length, int count)
        : base(name)
    {
        this.width = width;
        this.height = height;
        this.length = length;
        this.count = count;
    }

    /// <summary>
    /// Получить ширину
    /// </summary>
    public float GetWidth()
    {
        return width;
    }

    /// <summary>
    /// Получить высоту
    /// </summary>
    public float GetHeight()
    {
        return height;
    }

    /// <summary>
    /// Получить длину
    /// </summary>
    public float GetLength()
    {
        return length;
    }

    /// <summary>
    /// Получить количество
    /// </summary>
    public int GetCount()
    {
        return count;
    }

    /// <summary>
    /// Установить ширину
    /// </summary>
    public void SetWidth(float width)
    {
        this.width = width;
    }

    /// <summary>
    /// Установить высоту
    /// </summary>
    public void SetHeight(float height)
    {
        this.height = height;
    }

    /// <summary>
    /// Установить длину
    /// </summary>
    public void SetLength(float length)
    {
        this.length = length;
    }

    /// <summary>
    /// Установить количество
    /// </summary>
    public void SetCount(int count)
    {
        this.count = count;
    }

    /// <summary>
    /// Добавить количество
    /// </summary>
    public void AddCount(int amount)
    {
        count += amount;
        Console.WriteLine($"Added {amount} units. New count: {count}");
    }

    /// <summary>
    /// Удалить количество
    /// </summary>
    public void RemoveCount(int amount)
    {
        if (count >= amount)
        {
            count -= amount;
            Console.WriteLine($"Removed {amount} units. New count: {count}");
        }
        else
        {
            Console.WriteLine($"Cannot remove {amount} units. Only {count} available.");
        }
    }

    /// <summary>
    /// Вычислить объем одной единицы
    /// </summary>
    public float CalculateVolume()
    {
        return width * height * length;
    }

    /// <summary>
    /// Отображение информации о материале
    /// </summary>
    public override void DisplayInfo()
    {
        Console.WriteLine($"[Structural Material]");
        Console.WriteLine($"  Name: {name}");
        Console.WriteLine($"  Dimensions: {width}m x {height}m x {length}m");
        Console.WriteLine($"  Volume per unit: {CalculateVolume():F6} m3");
        Console.WriteLine($"  Count: {count} pieces");
    }
}

File: cpp/build-company/StructuralMaterial.cpp
#include "StructuralMaterial.hpp"
#include <iostream>
#include <iomanip>

// Реализация конструктора
StructuralMaterial::StructuralMaterial(const std::string& name, float width, float height, float length, int count)
    : Material(name), width(width), height(height), length(length), count(count) {
}

// Реализация деструктора
StructuralMaterial::~StructuralMaterial() {
    std::cout << "Вызван деструктор StructuralMaterial для: " << name << std::endl;
}

// Реализация геттеров
float StructuralMaterial::getWidth() const {
    return width;
}

float StructuralMaterial::getHeight() const {
    return height;
}

float StructuralMaterial::getLength() const {
    return length;
}

int StructuralMaterial::getCount() const {
    return count;
}

// Реализация сеттеров
void StructuralMaterial::setWidth(float width) {
    this->width = width;
}

void StructuralMaterial::setHeight(float height) {
    this->height = height;
}

void StructuralMaterial::setLength(float length) {
    this->length = length;
}

void StructuralMaterial::setCount(int count) {
    this->count = count;
}

// Добавление количества
void StructuralMaterial::addCount(int amount) {
    if (amount <= 0) return;
    count += amount;
    std::cout << "Добавлено " << amount << " шт. " << name
              << ". Новое количество: " << count << std::endl;
}

// Удаление количества
void StructuralMaterial::removeCount(int amount) {
    if (amount <= 0) return;
    if (amount > count) {
        std::cout << "Ошибка: Невозможно удалить " << amount << ". Доступно только " << count << std::endl;
        return;
    }
    count -= amount;
    std::cout << "Удалено " << amount << " шт. " << name
              << ". Осталось: " << count << std::endl;
}

// Вычисление объема одной единицы
float StructuralMaterial::calculateVolume() const {
    return width * height * length;
}

// Реализация функции отображения информации
void StructuralMaterial::displayInfo() const {
    std::cout << "=== Конструкционный материал ===" << std::endl;
    std::cout << "Название: " << name << std::endl;
    std::cout << "Размеры: " << std::fixed << std::setprecision(2)
              << width << " x " << height << " x " << length << " м" << std::endl;
    std::cout << "Количество: " << count << " шт." << std::endl;
    std::cout << "Объем единицы: " << std::fixed << std::setprecision(4)
              << calculateVolume() << " м3" << std::endl;
}
